cilium/CFP-32810-hybrid-routing-mode.md

```diff
@@ -0,0 +1,163 @@
+2. In a large Kubernetes cluster, pod addresses are assigned from two different subnets with no direct connection.
+
+In either case, Cilium needs to run in encapsulation routing mode. However, when routing packets in above clusters, in many situations, we can get by native routing, and not pay the encapsulation overhead that comes with encapsulation. The reason for this inflexibility comes from the fact that Cilium datapath is not aware of the underlying network topology, and hence cannot make a decision between the two routing modes during runtime.
+
+With higher adoption in cluster mesh and large scale clusters being common to support AI workloads, it would be beneficial to have a more performant dataplane that can optimize the throughput, while providing support for both encapsulation and native-routing.
+
+## Goals
+
+* Retain the behavior of current routing modes as-is
+* Allow users to set a new routing mode called `hybrid`
+* Allow users to pass in network topologies expressed as CIDR ranges through configmap
+* Support for both IPV4 and IPV6
+* Allow dynamic configuration of CIDRs (add/remove) without agent restart (update w/o agent restart)
+* Support for multiple subnets that are not not all mutually routable
+
+## Proposal
+
+### Overview
+
+Add a new option for routing-mode config as follows `routing-mode: hybrid`. Also, add a new field in config to pass information about the CIDRs and connectivity. The CIDR information will persist in pinned eBPF maps so that routing decision can be made in runtime by eBPF programs. The concept is similar to [IP Masquerading](https://docs.cilium.io/en/stable/network/concepts/masquerading/#ebpf-based) and the eBPF table it uses.
+
+### Examples
+
+```
+subnet-topology: 10.0.0.1/24
+```
+
+If source and destination IP belongs to above CIDR, route natively. Else, preserve current behavior (encapsulate if being routed to known CIDRs/pass to Linux stack).
+
+```
+subnet-topology: 10.0.0.1/24,10.10.0.1/24
+```
+
+The above two CIDR ranges have direct connectivity. If source and destination IP belongs to either CIDR, route natively. Else, preserve current behavior (encapsulate if being routed to know CIDRs/pass to Linux stack).
+```
+subnet-topology: 10.0.0.1/24,10.10.0.1/24;10.20.0.1/24
+```
+
+* `10.0.0.1/24` and `10.10.0.1/24` are connected and traffic can be natively routed
+* If source and destination IP belongs to `10.20.0.1/24`, route natively
+* Else, preserve current behavior (encapsulate if being routed to know CIDRs/pass to Linux stack).
+
+### Control plane implementation
+
+Similar to dynamic flowlogs or Hubble dynamic exporter, allow users to configure the CIDRs on the fly through a configmap. The configmap will be mounted on each agent. On create/update, update `LocalNodeConfig` and trigger the eBPF map reconcile.
+
+Here's how the different topologies would be represented in-memory cache for the following config: `10.0.0.1/24,10.10.0.1/24;10.20.0.1/24;2001:0db8:85a3::/64`
```

Comment:
why configmap and not CRD? Isn't this easy to represent in CRD than configmap?

Reply:
We discussed this here - https://github.com/cilium/design-cfps/pull/73#discussion_r2199013817 .
Possibility of adding a CRD is a future milestone.